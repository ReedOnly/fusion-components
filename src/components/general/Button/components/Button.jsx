import React, { forwardRef } from "react";
import Content from "./Content";
import getButtonClasses from "../buttonClasses";

const ButtonComponent = forwardRef(
    (
        {
            children,
            disabled,
            onMouseDown,
            onMouseUp,
            onClick,
            onClickCapture,
            ...props
        },
        ref
    ) => (
        <button
            type="button"
            className={getButtonClasses(props)}
            disabled={disabled}
            onMouseDown={onMouseDown}
            onMouseUp={onMouseUp}
            onClick={onClick}
            onClickCapture={onClickCapture}
            ref={ref}
        >
            <Content>{children}</Content>
        </button>
    )
);

ButtonComponent.displayName =
    "@fusion/components/general/Button/ButtonComponent";

export default ButtonComponent;
