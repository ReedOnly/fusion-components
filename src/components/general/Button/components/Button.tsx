import * as React from "react";
import Content from "./Content";
import getButtonClasses from "../buttonClasses";

type ButtonProps = {
    disabled?: boolean,
    onClick?: React.EventHandler<React.SyntheticEvent>,
    onClickCapture?: React.EventHandler<React.SyntheticEvent>,
    onMouseDown?: React.EventHandler<React.SyntheticEvent>,
    onMouseUp?: React.EventHandler<React.SyntheticEvent>,
    displayType: string,
    mouseHasBeenDown: boolean,
};

const ButtonComponent = React.forwardRef<HTMLButtonElement | null, React.PropsWithChildren<ButtonProps>>(
    ({ children, disabled, onMouseDown, onMouseUp, onClick, onClickCapture, ...props }, ref) => (
        <button
            type="button"
            className={getButtonClasses(props)}
            disabled={disabled}
            onMouseDown={onMouseDown}
            onMouseUp={onMouseUp}
            onClick={onClick}
            onClickCapture={onClickCapture}
            ref={ref}
        >
            <Content>{children}</Content>
        </button>
    )
);

ButtonComponent.defaultProps = {
    disabled: true,
    onClick: () => {},
    onClickCapture: () => {},
    onMouseDown: () => {},
    onMouseUp: () => {},
};

ButtonComponent.displayName = "@fusion/components/general/Button/ButtonComponent";

export default ButtonComponent;
