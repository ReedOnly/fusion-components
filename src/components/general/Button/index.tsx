import * as React from 'react';
import ComponentDisplayContext from '../../contexts/componentDisplayContext';
import ButtonComponent from './components/Button';
import AnchorComponent from './components/Anchor';

type ButtonProps = {
    disabled?: boolean;

    // Variants
    /** Use the contained variant */
    contained?: boolean;
    /** Use the outlined variant */
    outlined?: boolean;
    /** Use the frameless variant */
    frameless?: boolean;

    // Styles
    /** Use the primary coloring */
    primary?: boolean;
    /** Use the signal coloring */
    signal?: boolean;

    // Sizes
    /** Compact button */
    compact?: boolean;
    /** Comfortable button */
    comfortable?: boolean;

    // Link
    /** Provide an url and the button wil use <a>-tag instead of <button>-tag */
    url?: string | null;
    /** Set target="_blank" */
    targetBlank?: boolean;
    onMouseDown?: React.EventHandler<React.SyntheticEvent>;
    onClickCapture?: React.EventHandler<React.SyntheticEvent>;
    onClick?: React.EventHandler<React.SyntheticEvent>;
};

const Button = React.forwardRef<HTMLElement | null, React.PropsWithChildren<ButtonProps>>(
    (props, ref) => {
        // Used to apply "radar" animation on mouse up
        const [mouseHasBeenDown, setMouseHasBeenDown] = React.useState(false);
        const displayType = React.useContext<string>(ComponentDisplayContext);

        const buttonRef =
            (ref as React.RefObject<HTMLElement | null>) || React.createRef<HTMLElement | null>();

        const handleOnMouseDown = (e: React.SyntheticEvent) => {
            const { onMouseDown } = props;

            setMouseHasBeenDown(false);

            if (onMouseDown) {
                onMouseDown(e);
            }
        };

        const handleOnMouseUp = () => {
            setMouseHasBeenDown(true);
            // Focus is automatically set to the button on click,
            // but we don't want it to have focus after the click
            // action is performed. Only when tabbing to the button
            if (buttonRef.current !== null) {
                buttonRef.current.blur();
            }
        };

        return props.url ? (
            <AnchorComponent
                {...props}
                displayType={displayType}
                ref={buttonRef as React.RefObject<HTMLAnchorElement | null>}
                mouseHasBeenDown={mouseHasBeenDown}
                onMouseDown={handleOnMouseDown}
                onMouseUp={handleOnMouseUp}
                onClick={props.onClick}
                onClickCapture={props.onClickCapture}
                url={props.url}
            />
        ) : (
            <ButtonComponent
                {...props}
                displayType={displayType}
                ref={buttonRef as React.RefObject<HTMLButtonElement | null>}
                mouseHasBeenDown={mouseHasBeenDown}
                onMouseDown={handleOnMouseDown}
                onMouseUp={handleOnMouseUp}
                onClick={props.onClick}
                onClickCapture={props.onClickCapture}
            />
        );
    }
);

Button.displayName = 'Button';

Button.defaultProps = {
    disabled: false,
    contained: false,
    outlined: false,
    frameless: false,
    primary: true,
    signal: false,
    compact: false,
    comfortable: true,
    url: null,
    targetBlank: false,
    onMouseDown: () => {},
    onClickCapture: () => {},
    onClick: () => {},
};

export default Button;
