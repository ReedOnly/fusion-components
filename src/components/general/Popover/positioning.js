export const popoverHorizontalPositions = {
    left: "left",
    right: "right",
};

export const popoverVerticalPositions = {
    top: "top",
    bottom: "bottom",
};
