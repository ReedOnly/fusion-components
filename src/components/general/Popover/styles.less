@import "~styles/variables.less";

.container {
    position: relative;
    pointer-events: all;
    max-height: 75vh;
    display: flex;

    &.isRelative {
        position: absolute;
        top: 100%;
    }

    &.left {
        left: 0;

        .arrow {
            left: 11px;
        }
    }

    &.right {
        right: 0;

        .arrow {
            right: 11px;
        }
    }

    &.isInset {
        top: calc(100% - calc(var(--grid-unit) * 2px));

        @media (max-width: var(--mobile-max-width)) {
            top:100%;
        }

        &.left {
            left: calc(var(--grid-unit) * 3px);
        }

        &.right {
            right: calc(var(--grid-unit) * 1px);
        }
    }

    .content {
        background: white;
        overflow: hidden;
        border-radius: 4px;
        box-shadow: 0 0 4px rgba(0, 0, 0, .1), 0 4px 4px rgba(0, 0, 0, .05);
        display: flex;
        max-height: 100%;
        box-sizing: border-box;
        padding: calc(var(--grid-unit) * 1px)
    }

    .arrow {
        position: absolute;
        bottom: 100%;
        width: 10px;
        height: 4px;

        > svg {
            position: absolute;
            top: -4px;
            left: -4px;

            path {
                box-shadow: 0 0 4px rgba(0, 0, 0, 0.1), 0 4px 4px rgba(0, 0, 0, 0.05);
            }
        }
    }
}