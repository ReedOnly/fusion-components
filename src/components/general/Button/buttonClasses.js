import classNames from "classnames";
import { componentDisplayTypes } from "../../contexts/ComponentDisplayContext";
import styles from "./styles/index.less";

const getLayoutClasses = ({ block, flex }) => ({
    [styles.block]: block,
    [styles.flex]: flex,
});

const getMouseDownClasses = ({ mouseHasBeenDown }) => ({
    [styles.mouseHasBeenDown]: mouseHasBeenDown,
});

const getButtonVariantClasses = ({ contained, outlined, frameless }) => ({
    [styles.contained]: contained || (!outlined && !frameless), // Default to contained
    [styles.outlined]: outlined,
    [styles.frameless]: frameless,
    [styles.icon]: false,
});

const getButtonStyleClasses = ({ primary, signal }) => ({
    [styles.primary]: primary,
    [styles.signal]: signal,
});

const getButtonSizeClasses = ({ compact, comfortable, displayType }) => ({
    [styles.compact]: displayType === componentDisplayTypes.compact || compact,
    [styles.comfortable]:
        (displayType === componentDisplayTypes.comfortable &&
            displayType !== componentDisplayTypes.comfortable) ||
        (comfortable && !compact),
});

export default props =>
    classNames(
        props.className,
        styles.container,
        getLayoutClasses(props),
        getMouseDownClasses(props),
        getButtonVariantClasses(props),
        getButtonStyleClasses(props),
        getButtonSizeClasses(props)
    );
