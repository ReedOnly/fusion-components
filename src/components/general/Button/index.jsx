import React, { Component } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import ComponentDisplayContext, {
    componentDisplayTypes,
} from "../../contexts/ComponentDisplayContext";
import styles from "./styles.less";

class Button extends Component {
    static propTypes = {
        children: PropTypes.node.isRequired,
        /** Disable the button */
        disabled: PropTypes.bool,

        // Variants
        /** Use the contained variant */
        contained: PropTypes.bool,
        /** Use the outlined variant */
        outlined: PropTypes.bool,
        /** Use the frameless variant */
        frameless: PropTypes.bool,

        // Styles
        /** Use the primary coloring */
        primary: PropTypes.bool,
        /** Use the signal coloring */
        signal: PropTypes.bool,

        // Sizes
        /** Small button */
        small: PropTypes.bool,
        /** Medium button */
        medium: PropTypes.bool,
        /** Large button */
        large: PropTypes.bool,

        // Link
        /** Provide an url and the button wil use <a>-tag instead of <button>-tag */
        url: PropTypes.string,
        /** Set target="_blank" */
        targetBlank: PropTypes.bool,
        onMouseDown: PropTypes.func,
        onClickCapture: PropTypes.func,
        onClick: PropTypes.func,
    };

    static defaultProps = {
        disabled: false,
        contained: false,
        outlined: false,
        frameless: false,
        primary: true,
        signal: false,
        small: true,
        medium: true,
        large: false,
        url: null,
        targetBlank: false,
        onMouseDown: () => {},
        onClickCapture: () => {},
        onClick: () => {},
    };

    state = {
        isMouseDown: false,
        mouseHasBeenDown: false,
    };

    onMouseUp() {
        this.setState({
            isMouseDown: false,
            mouseHasBeenDown: true,
        });

        if (this.button) {
            this.button.blur();
        }
    }

    onMouseDown(e) {
        const { onMouseDown } = this.props;
        if (onMouseDown) {
            onMouseDown(e);
        }

        this.setState({
            isMouseDown: true,
            mouseHasBeenDown: false,
        });
    }

    getMouseDownClasses = () => {
        const { isMouseDown, mouseHasBeenDown } = this.state;

        return {
            [styles.mouseIsDown]: isMouseDown,
            [styles.mouseHasBeenDown]: mouseHasBeenDown,
        };
    };

    getButtonVariantClasses = () => {
        const { contained, outlined, frameless } = this.props;

        return {
            [styles.contained]: contained || (!outlined && !frameless), // Default to contained
            [styles.outlined]: outlined,
            [styles.frameless]: frameless,
            [styles.icon]: false,
        };
    };

    getButtonStyleClasses = () => {
        const { primary, signal } = this.props;
        return {
            [styles.primary]: primary,
            [styles.signal]: signal,
        };
    };

    getButtonSizeClasses = displayType => {
        const { small, medium, large } = this.props;

        return {
            [styles.small]:
                small || displayType === componentDisplayTypes.compact,
            [styles.medium]:
                (medium || (!small && !large)) &&
                displayType !== componentDisplayTypes.compact, // Default to medium
            [styles.large]:
                large && displayType !== componentDisplayTypes.compact,
        };
    };

    getButtonClasses = (displayType, props) =>
        classNames(
            props.className,
            styles.container,
            {
                [styles.block]: props.block,
                [styles.flex]: props.flex,
            },
            this.getMouseDownClasses(props),
            this.getButtonVariantClasses(props),
            this.getButtonStyleClasses(props),
            this.getButtonSizeClasses(displayType)
        );

    renderButtonContent() {
        const { children } = this.props;
        return <span className={styles.button}>{children}</span>;
    }

    renderButton(displayType) {
        const {
            disabled,
            onClick,
            onClickCapture,
            url,
            targetBlank,
        } = this.props;

        const containerClassNames = this.getButtonClasses(displayType, {
            ...this.props,
            ...this.state,
        });

        if (url) {
            return (
                <a
                    className={containerClassNames}
                    disabled={disabled}
                    href={url}
                    target={targetBlank ? "_blank" : "_self"}
                    onMouseDown={() => this.onMouseDown()}
                    onMouseUp={() => this.onMouseUp()}
                    onClickCapture={onClickCapture}
                    ref={button => {
                        this.button = button;
                    }}
                >
                    {this.renderButtonContent()}
                </a>
            );
        }
        return (
            <button
                type="button"
                className={containerClassNames}
                disabled={disabled}
                onClick={onClick}
                onMouseDown={e => this.onMouseDown(e)}
                onMouseUp={() => this.onMouseUp()}
                onClickCapture={onClickCapture}
                ref={button => {
                    this.button = button;
                }}
            >
                {this.renderButtonContent()}
            </button>
        );
    }

    render() {
        return (
            <ComponentDisplayContext.Consumer>
                {displayType => this.renderButton(displayType)}
            </ComponentDisplayContext.Consumer>
        );
    }
}

export default Button;
