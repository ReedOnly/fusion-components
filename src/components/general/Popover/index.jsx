import React, { Component } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import RelativePortal from "components/utils/RelativePortal";
import Arrow from "./Arrow";
import styles from "./styles.less";

export const popoverHorizontalPositions = {
    left: "left",
    right: "right",
};

export const popoverVerticalPositions = {
    top: "top",
    bottom: "bottom",
};

export default class Popover extends Component {
    static propTypes = {
        isOpen: PropTypes.bool.isRequired,
        onClose: PropTypes.func.isRequired,
        relativeTo: PropTypes.node,
        horizontalPosition: PropTypes.oneOf(
            Object.values(popoverHorizontalPositions)
        ),
        verticalPosition: PropTypes.oneOf(
            Object.values(popoverVerticalPositions)
        ),
        inset: PropTypes.bool,
        children: PropTypes.node.isRequired,
    };

    static defaultProps = {
        inset: false,
        relativeTo: null,
    };

    static defaultProps = {
        horizontalPosition: popoverHorizontalPositions.left,
        verticalPosition: popoverVerticalPositions.bottom,
    };

    handleClickOutside() {
        const { isOpen } = this.props;
        if (isOpen) {
            setTimeout(() => this.close());
        }
    }

    close() {
        const { onClose } = this.props;
        if (onClose) {
            onClose();
        }
    }

    renderPopover() {
        const {
            relativeTo,
            children,
            horizontalPosition,
            verticalPosition,
            inset,
        } = this.props;

        const containerClassNames = classNames(
            styles.container,
            styles[horizontalPosition],
            styles[verticalPosition],
            {
                [styles.isInset]: inset,
                [styles.isRelative]: relativeTo,
            }
        );

        return (
            <div className={containerClassNames}>
                <Arrow />
                <div className={styles.content}>{children}</div>
            </div>
        );
    }

    render() {
        const { isOpen, relativeTo } = this.props;

        if (!isOpen) {
            return null;
        }

        if (!relativeTo) {
            return this.renderPopover();
        }

        return (
            <RelativePortal
                relativeTo={relativeTo}
                onClickOutside={e => this.handleClickOutside(e)}
            >
                {this.renderPopover()}
            </RelativePortal>
        );
    }
}
