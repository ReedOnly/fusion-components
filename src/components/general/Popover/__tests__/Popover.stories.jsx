import React, { useState, createRef } from "react";
import { storiesOf } from "@storybook/react";
import { action } from "@storybook/addon-actions";
import Button from "components/general/Button";
import Popover from "..";

const buttonRef = createRef();
const ControlledPopover = () => {
    const [showPopover, setShowPopover] = useState(false);

    return (
        <React.Fragment>
            <Button
                primary
                contained
                onClick={() => setShowPopover(!showPopover)}
                ref={buttonRef}
            >
                Click me
            </Button>
            <Popover
                isOpen={showPopover}
                relativeTo={buttonRef.current}
                onClose={() => action("close")}
            >
                <div
                    style={{
                        padding: 8,
                        width: 400,
                        fontFamily: "Equinor",
                    }}
                >
                    <h2>Hello there!</h2>
                    <p>
                        Lorem ipsum dolor sit amet, consectetur adipiscing elit.
                        Fusce eu mi eu ipsum venenatis aliquam nec auctor ante.
                        Interdum et malesuada fames ac ante ipsum primis in
                        faucibus.
                    </p>
                    <div style={{ textAlign: "right" }}>
                        <Button
                            small
                            primary
                            frameless
                            onClick={() => setShowPopover(false)}
                        >
                            Close
                        </Button>
                    </div>
                </div>
            </Popover>
        </React.Fragment>
    );
};

storiesOf("General components/Popover", module)
    .addParameters({ jest: ["Popover.stories"] })
    .add("Inline", () => (
        <div>
            <Popover isOpen>
                <div style={{ padding: 8 }}>Some content</div>
            </Popover>
        </div>
    ))
    .add("Relative", () => <ControlledPopover />);
