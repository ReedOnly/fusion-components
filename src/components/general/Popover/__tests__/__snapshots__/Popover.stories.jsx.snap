// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`General components/Popover Inline 1`] = `
<div>
  <div
    className="container left bottom"
  >
    <div
      className="arrow"
    >
      <svg
        fill="none"
        height="16"
        viewBox="0 0 18 16"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          filter="url(#filter0_dd)"
        >
          <path
            d="M9 4L4 8L14 8L9 4Z"
            fill="white"
          />
        </g>
        <defs>
          <filter
            colorInterpolationFilters="sRGB"
            filterUnits="userSpaceOnUse"
            height="16"
            id="filter0_dd"
            width="18"
            x="0"
            y="0"
          >
            <feFlood
              floodOpacity="0"
              result="BackgroundImageFix"
            />
            <feColorMatrix
              in="SourceAlpha"
              type="matrix"
              values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0"
            />
            <feOffset
              dy="-1.7"
            />
            <feGaussianBlur
              stdDeviation="1.7"
            />
            <feColorMatrix
              type="matrix"
              values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.04 0"
            />
            <feBlend
              in2="effect1_dropShadow"
              mode="normal"
              result="effect2_dropShadow"
            />
            <feBlend
              in="SourceGraphic"
              in2="effect2_dropShadow"
              mode="normal"
              result="shape"
            />
          </filter>
        </defs>
      </svg>
    </div>
    <div
      className="content"
    >
      <div
        style={
          Object {
            "padding": 8,
          }
        }
      >
        Some content
      </div>
    </div>
  </div>
</div>
`;

exports[`General components/Popover Relative 1`] = `
<button
  className="container contained primary small medium"
  disabled={false}
  onClick={[Function]}
  onClickCapture={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  type="button"
>
  <span
    className="button"
  >
    Click me
  </span>
</button>
`;
